client/snapshot.go

```diff
@@ -0,0 +1,174 @@
+	SetID  uint64   `json:"set"`
+	Action string   `json:"action"`
+	Snaps  []string `json:"snaps,omitempty"`
+	Users  []string `json:"users,omitempty"`
+}
+
+// A Snapshot is a collection of archives with a simple metadata json file
+// (and hashsums of everything).
+type Snapshot struct {
+	// SetID is the ID of the snapshot set (a snapshot set is a "snap save" invocation)
+	SetID uint64 `json:"set"`
+	// the time this snapshot's data collection was started
+	Time time.Time `json:"time"`
+
+	// information about the snap this data is for
+	Snap     string        `json:"snap"`
+	Revision snap.Revision `json:"revision"`
+	Epoch    snap.Epoch    `json:"epoch,omitempty"`
+	Summary  string        `json:"summary"`
+	Version  string        `json:"version"`
+
+	// the snap's configuration at snapshot time
+	Conf map[string]interface{} `json:"conf,omitempty"`
+
+	// the hash of the archives' data, keyed by archive path
+	// (either 'archive.tgz' for the system archive, or
+	// user/<username>.tgz for each user)
+	SHA3_384 map[string]string `json:"sha3-384"`
+	// the sum of the archive sizes
+	Size int64 `json:"size,omitempty"`
+	// if the snapshot failed to open this will be the reason why
+	Broken string `json:"broken,omitempty"`
+}
+
+// IsValid checks whether the snapshot is missing information that
+// should be there for a snapshot that's just been opened.
+func (sh *Snapshot) IsValid() bool {
+	return !(sh == nil || sh.SetID == 0 || sh.Snap == "" || sh.Revision.Unset() || len(sh.SHA3_384) == 0 || sh.Time.IsZero())
+}
+
+// A SnapshotSet is a set of snapshots created by a single "snap save".
+type SnapshotSet struct {
+	ID        uint64      `json:"id"`
+	Snapshots []*Snapshot `json:"snapshots"`
+}
+
+// Time returns the earliest time in the set.
+func (ss SnapshotSet) Time() time.Time {
+	if len(ss.Snapshots) == 0 {
+		return time.Time{}
+	}
+	mint := ss.Snapshots[0].Time
+	for _, sh := range ss.Snapshots {
+		if sh.Time.Before(mint) {
+			mint = sh.Time
+		}
+	}
+	return mint
+}
+
+// Size returns the sum of the set's sizes.
+func (ss SnapshotSet) Size() int64 {
+	var sum int64
+	for _, sh := range ss.Snapshots {
+		sum += sh.Size
+	}
+	return sum
+}
+
+// SnapshotSets lists the snapshot sets in the system that belong to the
+// given set (if non-zero) and are for the given snaps (if non-empty).
+func (client *Client) SnapshotSets(setID uint64, snapNames []string) ([]SnapshotSet, error) {
+	q := make(url.Values)
+	if setID > 0 {
+		q.Add("set", strconv.FormatUint(setID, 10))
+	}
+	if len(snapNames) > 0 {
+		q.Add("snaps", strings.Join(snapNames, ","))
+	}
+
+	var snapshotSets []SnapshotSet
+	_, err := client.doSync("GET", "/v2/snapshots", q, nil, nil, &snapshotSets)
+	return snapshotSets, err
+}
+
+// ForgetSnapshots permanently removes the snapshot set, limited to the
+// given snaps (if non-empty).
+func (client *Client) ForgetSnapshots(setID uint64, snaps []string) (changeID string, err error) {
+	return client.snapshotAction(&snapshotAction{
```

Comment:
do you need to check if `setID != 0`? (probably the same in CheckSnapshots and RestoreSapshots)

Reply:
does it add anything? I mean, if you call it with `setID == 0` you'll get back an `ErrSnapshotSetNotFound` which should be clear enough.